validate.go

```diff
@@ -0,0 +1,88 @@
+package vat
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// Validate validates a VAT number by both format and existence. If no error then it is valid.
+func Validate(vatNumber string) error {
+	err := ValidateFormat(vatNumber)
+	if err != nil {
+		return err
+	}
+	return ValidateExists(vatNumber)
+}
+
+// ValidateFormat validates a VAT number by its format. If no error is returned then it is valid.
+func ValidateFormat(vatNumber string) error {
+	patterns := map[string]string{
+		"AT": `U[A-Z0-9]{8}`,
+		"BE": `(0[0-9]{9}|[0-9]{10})`,
+		"BG": `[0-9]{9,10}`,
+		"CH": `(?:E(?:-| )[0-9]{3}(?:\.| )[0-9]{3}(?:\.| )[0-9]{3}( MWST)?|E[0-9]{9}(?:MWST)?)`,
+		"CY": `[0-9]{8}[A-Z]`,
+		"CZ": `[0-9]{8,10}`,
+		"DE": `[0-9]{9}`,
+		"DK": `[0-9]{8}`,
+		"EE": `[0-9]{9}`,
+		"EL": `[0-9]{9}`,
+		"ES": `[A-Z][0-9]{7}[A-Z]|[0-9]{8}[A-Z]|[A-Z][0-9]{8}`,
+		"FI": `[0-9]{8}`,
+		"FR": `([A-Z]{2}|[0-9]{2})[0-9]{9}`,
+		// Supposedly the regex for GB numbers is `[0-9]{9}|[0-9]{12}|(GD|HA)[0-9]{3}`,
+		// but our validator service only accepts numbers with 9 or 12 digits following the country code.
+		// Seems like the official site only accepts 9 digits... https://www.gov.uk/check-uk-vat-number
+		"GB": `([0-9]{9}|[0-9]{12})`,
+		"HR": `[0-9]{11}`,
+		"HU": `[0-9]{8}`,
+		"IE": `[A-Z0-9]{7}[A-Z]|[A-Z0-9]{7}[A-W][A-I]`,
+		"IT": `[0-9]{11}`,
+		"LT": `([0-9]{9}|[0-9]{12})`,
+		"LU": `[0-9]{8}`,
+		"LV": `[0-9]{11}`,
+		"MT": `[0-9]{8}`,
+		"NL": `[0-9]{9}B[0-9]{2}`,
+		"PL": `[0-9]{10}`,
+		"PT": `[0-9]{9}`,
+		"RO": `[0-9]{2,10}`,
+		"SE": `[0-9]{12}`,
+		"SI": `[0-9]{8}`,
+		"SK": `[0-9]{10}`,
+		"XI": `([0-9]{9}|[0-9]{12})`, // Northern Ireland, same format as GB
+	}
+
+	if len(vatNumber) < 3 {
+		return ErrInvalidVATNumberFormat
+	}
+
+	vatNumber = strings.ToUpper(vatNumber)
+	pattern, ok := patterns[vatNumber[0:2]]
+	if !ok {
+		return ErrInvalidCountryCode
+	}
+
+	matched, err := regexp.MatchString(fmt.Sprintf("^%s$", pattern), vatNumber[2:])
+	if err != nil {
+		return ErrUnexpected{Err: err}
```

Comment:
Personally I'd ditch the `ErrUnexpected` part and return `ErrInvalidVATNumberFormat` since the outcome is the same whether `regexp` itself encounters an error or not.

Reply:
But we don't expect this error to ever happen, do we? If it does it would be unexpected?

Like if there's no match, it just returns `false` and no error, right? It only would return an error if for example our regex is messed up or something like that we don't expect if I'm understanding the docs correctly:

https://pkg.go.dev/regexp#example-MatchString

Reply:
Looks like regexp will return false if an error occurs too. Either way we were unable to validate the number, which why I think the outcome is the same. Like you said, it mostly comes down to our own predefined expressions being the source of a possible error.

You can return the error too I suppose but maybe not wrap it in `ErrUnexpected` and just return the raw value. Sort of depends on what you want the caller of `vat` to do in this case.

Reply:
K just gonna unwrap the error in that case